packages/xl-ai/src/prosemirror/agent.ts

```diff
@@ -0,0 +1,325 @@
+      prosemirrorSteps: [],
+      selection: {
+        anchor: tr.mapping.map(step.from),
+        head: tr.mapping.map(step.to),
+      },
+      type: "select",
+    });
+
+    // Replace the content
+    const sliceTextContent = step.slice.content.textBetween(0, step.slice.size);
+
+    const alreadyHasSameText =
+      sliceTextContent ===
+      tr.doc.textBetween(tr.mapping.map(step.from), tr.mapping.map(step.to));
+
+    let sliceTo: number;
+
+    if (alreadyHasSameText) {
+      sliceTo = step.slice.content.size; // replace all at once, it's probably a mark update
+    } else if (sliceTextContent.length === 0) {
+      sliceTo = step.slice.content.size; // there's no replacement text, so use entire slice as replacement
+    } else {
+      // replace with the first character (similar to how a user would do it when selecting text and starting to type)
+      const firstCharIndex = getFirstChar(step.slice.content);
+      if (firstCharIndex === undefined) {
+        // should have been caught by previous if statement
+        throw new Error("unexpected: no first character found");
+      }
+      sliceTo = firstCharIndex + 1;
+    }
+
+    let replaceEnd = tr.mapping.map(step.to);
+    const replaceFrom = tr.mapping.map(step.to);
+    let first = true;
+
+    // Note that below, we don't actually delete content, but we mark it as deleted
+    // with a deletion mark. Similarly, we mark inserted content with an insertion mark.
+    // It might be cleaner to;
+    // a) make this optional
+    // b) actually delete / insert the content and let prosemirror-suggest-changes handle the marks
+    for (let i = sliceTo; i <= step.slice.content.size; i++) {
+      // in the first agent step (first character replacement) we mark existing content (if any) as deleted
+      // (step.from !== step.to is used to check for replacements vs inserts)
+      const isReplacing = first && step.from !== step.to;
+
+      const stepIndex = tr.steps.length;
+      if (isReplacing) {
+        const $pos = tr.doc.resolve(tr.mapping.map(step.from));
+        if ($pos.nodeAfter?.isBlock) {
+          // mark the entire node as deleted. This can be needed for inline nodes or table cells
+          tr.addNodeMark($pos.pos, pmSchema.mark("deletion", {}));
+        }
+        tr.addMark($pos.pos, replaceEnd, pmSchema.mark("deletion", {}));
+        replaceEnd = tr.mapping.map(step.to);
+      }
+
+      // note, instead of inserting one charachter at a time at the end (a, b, c)
+      // we're replacing the entire part every time (a, ab, abc)
+      // would be cleaner to do just insertions, but didn't get this to work with the add operation
+      // (and this kept code relatively simple)
+      const replacement = new Slice(step.slice.content.cut(0, i), 0, 0);
+
+      tr.replace(replaceFrom, replaceEnd, replacement).addMark(
+        replaceFrom,
+        replaceFrom + replacement.content.size,
+        pmSchema.mark("insertion", {}),
+      );
```

Comment:
This will also affect how undo/redo stack works, since an addMark step cannot be merged with a replace step.

What I've done before for this, is to add the mark to the `replacement` content separately, so it doesn't need a separate addMark step